### PR TITLE
Sequence modifications

### PR DESCRIPTION
**Pull request by ritch**
- Simplify `authentication.user` binding / support rebinding user
 - Additional ACL step `authorizeResult`
 - Support ACLs that operate on model IDs (via `args`)

### CODE REVIEW

**Comment by ritch** [comment]
@bajtos @superkhau @raymondfeng

File: ts/src/application.ts [hunk 1]
```diff
@@ -26,30 +27,34 @@ class TodoSequence extends Sequence {
...
+      // 2. Authenticate and bind current user
       const user = await this.authenticate(request);
-      
-      // 3. Authorize the operation
-      // TODO(bajtos) how to pass target model id for owner-based ACLs?
-      await this.authorize(controllerName, methodName, user);
+
+      // Anything produced / determined during the sequence
+      // should be bound into the request context
+      this.context.bind('authentication.user').to(user);
```
Comment (bajtos) [inline ts/src/application.ts:37]: As I have envisioned the sequence steps, and also how the current implementation of `findRoute` works, these bindings should be created by the sequence steps. See https://github.com/strongloop/loopback-next/blob/67841d32131b896e8b09cb0b4e171f8ca067bab3/packages/core/src/http-handler.ts#L54-L72

Not a big deal, your proposal is fine too, and we can tune this incrementally.

From what I have observed while working on the initial Sequence implementation, there are certain implementation constraints than have to be accounted for in the design of the acceptance tests. I find it difficult to foresee them in advance, before there is at least a partial implementation to back the design. In that light, I am proposing to move the acceptance tests back into loopback-next repository and include changes in the implementation when changing acceptance tests.

File: ts/src/application.ts [hunk 2]
```diff
@@ -26,30 +27,34 @@ class TodoSequence extends Sequence {
...
-      // I would do the following:
-      //    const result = await this.invoke(controllerName, methodName, args, user);
       const result = await this.invoke(controllerName, methodName, args);
-      
-      // 6. Serialize the result (typically to JSON) and write HTTP response data
+
+      // 6. Authorize the result should be sent to the user
+      // These access controls operate on normal access meta as well
+      // as the result
+      await this.authorizeResult(controllerName, methodName, user, args, result);
```
Comment (bajtos) [inline ts/src/application.ts:55]: I don't fully understand this. What kind of authorization do you envision to happen on the result?

The first thing that comes to my mind is to filter out some data from the response depending on what the user is authorized to see. If that's the case, then I am concerned about performance implications of fetching unnecessary data from backend services, and also about security implications - e.g. if `authorizeResult` is implemented via a black-list of what should be removed, then such implementation can easily miss new data/properties/records that are added by future changes of the application.

Comment (raymondfeng) [reply ts/src/application.ts:55]: Removing unauthorized data from the response is the probably the main use case. Some of the ACL rules can be  enforced by mutating the query, for example, adding `fields` or `where`. But I'm sure if that's always possible.

**Comment by bajtos** [comment]
Closing as no longer relevant.

**Pull request closed without merging**